### PR TITLE
[Merged by Bors] - feat: add List.enum_eq_nil and List.mapIdx_eq_nil

### PR DESCRIPTION
**Pull request by dwrensha**
Adds `List.enum_eq_nil` and `List.mapIdx_eq_nil`.
These are analogous to [`List.map_eq_nil`](https://github.com/leanprover/std4/blob/c14f6a65b2c08caa38e1ab5db83451460d6cde3e//Std/Data/List/Lemmas.lean#L166-L167).

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by dwrensha** [comment]
I ran into the need for these lemmas here: https://github.com/dwrensha/compfiles/blob/4e9c73e01a7a1633e7d88cbba8b4afc542aa251d/Compfiles/Imo1968P2.lean#L25-L31

**Comment by alexjbest** [comment]
maintainer merge

File: Mathlib/Data/List/Basic.lean [hunk 1]
```diff
@@ -3898,6 +3898,10 @@ theorem nthLe_enum (l : List α) (i : ℕ) (hi' : i < l.enum.length)
     l.enum.nthLe i hi' = (i, l.nthLe i hi) := get_enum ..
 #align list.nth_le_enum List.nthLe_enum
 
+@[simp]
+theorem enum_eq_nil {l : List α} : List.enum l = [] ↔ l = [] := by
+  cases l <;> simp [List.enum_cons]
```
Comment (eric-wieser) [inline Mathlib/Data/List/Basic.lean:3903]: The other `enum` theorems are proved in terms of an `enumFrom` lemma. Can you write that lemma for this case too?

**Comment by eric-wieser** [comment]
bors d+

**Pull request closed without merging**